aten/src/ATen/native/xnnpack/Convolution.cpp

```diff
@@ -0,0 +1,309 @@
+  std::vector<int64_t> padding;
+  std::vector<int64_t> stride;
+  std::vector<int64_t> dilation;
+
+  static constexpr float kMin = -std::numeric_limits<float>::infinity();
+  static constexpr float kMax = std::numeric_limits<float>::infinity();
+};
+
+namespace {
+
+// Supports NHWC and NCHW FP32 convolutions with any valid
+//  - kernel size
+//  - padding
+//  - stride
+//  - dilation
+//  - grouping
+
+bool available(
+    const Tensor& weight,
+    const c10::optional<Tensor>& bias,
+    const IntArrayRef padding,
+    const IntArrayRef stride,
+    const IntArrayRef dilation,
+    const int64_t groups,
+    const float output_min,
+    const float output_max) {
+         // XNNPACK
+  return xnnpack::internal::available() &&
+         // Weight
+         (4 == weight.ndimension()) &&
+         (weight.size(Layout::Filter::Height) > 0) &&
+         (weight.size(Layout::Filter::Width) > 0) &&
+         (c10::DeviceType::CPU == weight.device().type()) &&
+         (kFloat == weight.scalar_type()) &&
+         // Bias
+         ((bias && bias->defined()) ? ((1 == bias->ndimension()) &&
+                                      (c10::DeviceType::CPU == bias->device().type()) &&
+                                      (kFloat == bias->scalar_type()) &&
+                                      (weight.size(Layout::Filter::Output)) == bias->size(0))
+                                    : true) &&
+         // Padding
+         (padding[Layout::Parameter::Height] >= 0) &&
+         (padding[Layout::Parameter::Width] >= 0) &&
+         // Stride
+         (stride[Layout::Parameter::Height] > 0) &&
+         (stride[Layout::Parameter::Width] > 0) &&
+         // Dilation
+         (dilation[Layout::Parameter::Height] > 0) &&
+         (dilation[Layout::Parameter::Width] > 0) &&
+         // Groups
+         (groups > 0) &&
+         // Input
+         (weight.size(Layout::Filter::Input) > 0) &&
+         // Output
+         (weight.size(Layout::Filter::Output) > 0) &&
+         // Output - Groups
+         ((weight.size(Layout::Filter::Output) % groups) == 0) &&
+         // Output Min / Max
+         (output_max > output_min);
+}
+
+Context create(
+    const Tensor& weight,
+    const c10::optional<Tensor>& bias,
+    const IntArrayRef padding_,
+    const IntArrayRef stride_,
+    const IntArrayRef dilation_,
+    const int64_t groups,
+    const float output_min,
+    const float output_max) {
+  const auto padding = expand_param_if_needed(padding_, "padding", 2);
+  const auto stride = expand_param_if_needed(stride_, "stride", 2);
+  const auto dilation = expand_param_if_needed(dilation_, "dilation", 2);
+  const Tensor weight_nhwc = weight.contiguous(MemoryFormat::ChannelsLast);
+
+  TORCH_CHECK(
+      available(
+          weight_nhwc,
+          bias,
+          padding,
+          stride,
+          dilation,
+          groups,
+          output_min,
+          output_max),
+      "xnnpack::convolution not available!");
```

Comment:
I'm a bit worried that this error message doesn't actually tell the user what the problem was.  Let's be sure to improve this if anyone gets confused.

Reply:
My idea was that if this error prints the line and file the user can investigate.  Otherwise if I want to make the error message super descriptive, which is also a possibility, I have to break that function into its constituent tests.  Is that what you have in mind?

Reply:
Yes.  Don't need to do it right now, though.